Error injecting beans in test sources

Having a bean defined in test sources that depends on the EM:

```java
@ApplicationScoped
public class TestDataFactory {

    @Inject
    protected EntityManager em;
}
```

and trying to inject it into a @QuarkusTest I got this Exception:

[io.quarkus.test.junit.QuarkusTestExtension] failed to instantiate test class [net.lucamasini.testcase.QuarkusResourceTest]: Failed to inject field protected net.lucamasini.testcase.TestDataFactory net.lucamasini.testcase.QuarkusResourceTest.dataFactory
Caused by: java.lang.RuntimeException: Failed to inject field protected net.lucamasini.testcase.TestDataFactory net.lucamasini.testcase.QuarkusResourceTest.dataFactory
Caused by: java.lang.NullPointerException: Managed Bean [class net.lucamasini.testcase.TestDataFactory] is null

Attached a test case.
[test-case.zip](https://github.com/quarkusio/quarkus/files/2998519/test-case.zip)

--- a/core/deployment/src/main/java/io/quarkus/deployment/builditem/TestClassPredicateBuildItem.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/builditem/TestClassPredicateBuildItem.java
@@ -1,0 +1,23 @@
+package io.quarkus.deployment.builditem;
+
+import java.util.function.Predicate;
+
+import org.jboss.builder.item.SimpleBuildItem;
+
+/**
+ * This is an optional build item that allows extensions to distinguish test classes from application classes. It is only
+ * available during tests.
+ */
+public final class TestClassPredicateBuildItem extends SimpleBuildItem {
+
+    private final Predicate<String> predicate;
+
+    public TestClassPredicateBuildItem(Predicate<String> predicate) {
+        this.predicate = predicate;
+    }
+
+    public Predicate<String> getPredicate() {
+        return predicate;
+    }
+
+}

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcAnnotationProcessor.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcAnnotationProcessor.java
@@ -23,6 +23,7 @@ import java.util.Collection;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.Set;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
@@ -44,6 +45,7 @@ import io.quarkus.arc.deployment.UnremovableBeanBuildItem.BeanClassNameExclusion
 import io.quarkus.arc.processor.AnnotationsTransformer;
 import io.quarkus.arc.processor.BeanDefiningAnnotation;
 import io.quarkus.arc.processor.BeanDeployment;
+import io.quarkus.arc.processor.BeanInfo;
 import io.quarkus.arc.processor.BeanProcessor;
 import io.quarkus.arc.processor.BeanProcessor.Builder;
 import io.quarkus.arc.processor.ReflectionRegistration;
@@ -61,6 +63,7 @@ import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
 import io.quarkus.deployment.builditem.GeneratedResourceBuildItem;
 import io.quarkus.deployment.builditem.ShutdownContextBuildItem;
+import io.quarkus.deployment.builditem.TestClassPredicateBuildItem;
 import io.quarkus.deployment.builditem.substrate.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.substrate.ReflectiveFieldBuildItem;
 import io.quarkus.deployment.builditem.substrate.ReflectiveMethodBuildItem;
@@ -109,6 +112,9 @@ public class ArcAnnotationProcessor {
 
     @Inject
     List<UnremovableBeanBuildItem> removalExclusions;
+
+    @Inject
+    Optional<TestClassPredicateBuildItem> testClassPredicate;
 
     /**
      * The configuration for ArC, the CDI-based injection facility.
@@ -231,6 +237,14 @@ public class ArcAnnotationProcessor {
         }
         for (UnremovableBeanBuildItem exclusion : removalExclusions) {
             builder.addRemovalExclusion(exclusion.getPredicate());
+        }
+        if (testClassPredicate.isPresent()) {
+            builder.addRemovalExclusion(new Predicate<BeanInfo>() {
+                @Override
+                public boolean test(BeanInfo bean) {
+                    return testClassPredicate.get().getPredicate().test(bean.getBeanClass().toString());
+                }
+            });
         }
 
         BeanProcessor beanProcessor = builder.build();

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcTestResourceProvider.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcTestResourceProvider.java
@@ -30,7 +30,8 @@ public class ArcTestResourceProvider implements TestResourceProvider {
                                 qualifiers.add(a);
                             }
                         }
-                        Object instance = container.instance(f.getGenericType(), qualifiers.toArray(EMPTY_ANNOTATION_ARRAY)).get();
+                        Object instance = container.instance(f.getGenericType(), qualifiers.toArray(EMPTY_ANNOTATION_ARRAY))
+                                .get();
                         f.setAccessible(true);
                         try {
                             f.set(test, instance);

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcTestResourceProvider.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcTestResourceProvider.java
@@ -4,38 +4,33 @@ import java.lang.annotation.Annotation;
 import java.lang.reflect.Field;
 import java.util.ArrayList;
 import java.util.List;
-import java.util.Set;
 
-import javax.enterprise.context.spi.CreationalContext;
-import javax.enterprise.inject.spi.Bean;
-import javax.enterprise.inject.spi.BeanManager;
 import javax.inject.Inject;
 import javax.inject.Qualifier;
 
 import io.quarkus.arc.Arc;
+import io.quarkus.arc.ArcContainer;
 import io.quarkus.deployment.test.TestResourceProvider;
 
 public class ArcTestResourceProvider implements TestResourceProvider {
 
+    public static final Annotation[] EMPTY_ANNOTATION_ARRAY = new Annotation[0];
+
     @Override
     public void inject(Object test) {
         Class<?> c = test.getClass();
+        ArcContainer container = Arc.container();
         while (c != Object.class) {
             for (Field f : c.getDeclaredFields()) {
                 if (f.isAnnotationPresent(Inject.class)) {
                     try {
-                        BeanManager beanManager = Arc.container().beanManager();
                         List<Annotation> qualifiers = new ArrayList<>();
                         for (Annotation a : f.getAnnotations()) {
                             if (a.annotationType().isAnnotationPresent(Qualifier.class)) {
                                 qualifiers.add(a);
                             }
                         }
-                        Set<Bean<?>> beans = beanManager.getBeans(f.getType(),
-                                qualifiers.toArray(new Annotation[qualifiers.size()]));
-                        Bean<?> bean = beanManager.resolve(beans);
-                        CreationalContext<?> ctx = beanManager.createCreationalContext(bean);
-                        Object instance = beanManager.getReference(bean, f.getType(), ctx);
+                        Object instance = container.instance(f.getGenericType(), qualifiers.toArray(EMPTY_ANNOTATION_ARRAY)).get();
                         f.setAccessible(true);
                         try {
                             f.set(test, instance);

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/BeanArchiveProcessor.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/BeanArchiveProcessor.java
@@ -75,7 +75,6 @@ public class BeanArchiveProcessor {
             additionalBeans.addAll(i.getBeanClasses());
         }
         additionalBeans.add(LifecycleEventRunner.class.getName());
-        additionalBeans.add(ActivateRequestContextInterceptor.class.getName());
         Set<DotName> additionalIndex = new HashSet<>();
         for (String beanClass : additionalBeans) {
             IndexingUtil.indexClass(beanClass, additionalBeanIndexer, applicationIndex, additionalIndex,

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanArchives.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanArchives.java
@@ -1,6 +1,7 @@
 package io.quarkus.arc.processor;
 
 import io.quarkus.arc.ActivateRequestContextInterceptor;
+import io.quarkus.arc.InjectableRequestContextController;
 import java.io.IOException;
 import java.io.InputStream;
 import java.lang.reflect.Modifier;
@@ -39,23 +40,9 @@ public final class BeanArchives {
     public static IndexView buildBeanArchiveIndex(IndexView... applicationIndexes) {
         List<IndexView> indexes = new ArrayList<>();
         Collections.addAll(indexes, applicationIndexes);
-        // Add CDI API builtin annotations if needed
-        if (!contains(DotNames.ANY, applicationIndexes)) {
-            indexes.add(buildCdiApiIndex());
-        }
-        if (!contains(DotName.createSimple(ActivateRequestContextInterceptor.class.getName()), applicationIndexes)) {
-            indexes.add(buildBuiltinIndex());
-        }
+        indexes.add(buildCdiApiIndex());
+        indexes.add(buildBuiltinIndex());
         return new IndexWrapper(CompositeIndex.create(indexes));
-    }
-
-    private static boolean contains(DotName className, IndexView... indexes) {
-        for (IndexView index : indexes) {
-            if (index.getClassByName(className) != null) {
-                return true;
-            }
-        }
-        return false;
     }
 
     private static IndexView buildCdiApiIndex() {
@@ -73,6 +60,7 @@ public final class BeanArchives {
     private static IndexView buildBuiltinIndex() {
         Indexer indexer = new Indexer();
         index(indexer, ActivateRequestContextInterceptor.class.getName());
+        index(indexer, InjectableRequestContextController.class.getName());
         return indexer.complete();
     }
 

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java
@@ -183,7 +183,10 @@ public class ClientProxyGenerator extends AbstractGenerator {
         ResultHandle context = creator.invokeInterfaceMethod(MethodDescriptors.ARC_CONTAINER_GET_ACTIVE_CONTEXT,
                 container, scope);
         BytecodeCreator inactiveBranch = creator.ifNull(context).trueBranch();
-        inactiveBranch.throwException(ContextNotActiveException.class, "");
+        ResultHandle exception = inactiveBranch.newInstance(
+                MethodDescriptor.ofConstructor(ContextNotActiveException.class, String.class),
+                inactiveBranch.invokeVirtualMethod(MethodDescriptors.OBJECT_TO_STRING, scope));
+        inactiveBranch.throwException(exception);
         AssignableResultHandle ret = creator.createVariable(Object.class);
         creator.assign(ret, creator.invokeInterfaceMethod(MethodDescriptors.CONTEXT_GET_IF_PRESENT, context, bean));
         BytecodeCreator isNullBranch = creator.ifNull(ret).trueBranch();

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/MethodDescriptors.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/MethodDescriptors.java
@@ -71,6 +71,8 @@ final class MethodDescriptors {
     static final MethodDescriptor OBJECT_EQUALS = MethodDescriptor.ofMethod(Object.class, "equals", boolean.class,
             Object.class);
 
+    static final MethodDescriptor OBJECT_TO_STRING = MethodDescriptor.ofMethod(Object.class, "toString", String.class);
+
     static final MethodDescriptor OBJECT_CONSTRUCTOR = MethodDescriptor.ofConstructor(Object.class);
 
     static final MethodDescriptor INTERCEPTOR_INVOCATION_POST_CONSTRUCT = MethodDescriptor.ofMethod(InterceptorInvocation.class,

--- a/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java
+++ b/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java
@@ -17,6 +17,7 @@
 package io.quarkus.arc;
 
 import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
 import java.util.Set;
 import java.util.function.Supplier;
 import javax.enterprise.context.ContextNotActiveException;
@@ -66,6 +67,16 @@ public interface ArcContainer {
      * @return a new instance handle
      */
     <T> InstanceHandle<T> instance(TypeLiteral<T> type, Annotation... qualifiers);
+    
+    /**
+     * Never returns null. However, the handle is empty if no bean matches/multiple beans match the specified type and
+     * qualifiers.
+     *
+     * @param type
+     * @param qualifiers
+     * @return a new instance handle
+     */
+    <X> InstanceHandle<X> instance(Type type, Annotation... qualifiers);
 
     /**
      * Never returns null. However, the handle is empty if no bean matches/multiple beans match the specified name.

--- a/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java
+++ b/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java
@@ -67,7 +67,7 @@ public interface ArcContainer {
      * @return a new instance handle
      */
     <T> InstanceHandle<T> instance(TypeLiteral<T> type, Annotation... qualifiers);
-    
+
     /**
      * Never returns null. However, the handle is empty if no bean matches/multiple beans match the specified type and
      * qualifiers.

--- a/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainerImpl.java
+++ b/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainerImpl.java
@@ -182,6 +182,12 @@ class ArcContainerImpl implements ArcContainer {
     }
 
     @Override
+    public <X> InstanceHandle<X> instance(Type type, Annotation... qualifiers) {
+        requireRunning();
+        return instanceHandle(type, qualifiers);
+    }
+
+    @Override
     public <T> Supplier<InstanceHandle<T>> instanceSupplier(Class<T> type, Annotation... qualifiers) {
         requireRunning();
 

--- a/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InjectableRequestContextController.java
+++ b/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InjectableRequestContextController.java
@@ -1,0 +1,40 @@
+package io.quarkus.arc;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.enterprise.context.ContextNotActiveException;
+import javax.enterprise.context.Dependent;
+import javax.enterprise.context.RequestScoped;
+import javax.enterprise.context.control.RequestContextController;
+
+@Dependent
+public class InjectableRequestContextController implements RequestContextController {
+
+    private final ManagedContext requestContext;
+    private final AtomicBoolean isActivator;
+
+    public InjectableRequestContextController() {
+        this.requestContext = Arc.container().requestContext();
+        this.isActivator = new AtomicBoolean(false);
+    }
+
+    @Override
+    public boolean activate() {
+        if (Arc.container().getActiveContext(RequestScoped.class) != null) {
+            return false;
+        }
+        requestContext.activate();
+        isActivator.set(true);
+        return true;
+    }
+
+    @Override
+    public void deactivate() throws ContextNotActiveException {
+        if (Arc.container().getActiveContext(RequestScoped.class) == null) {
+            throw new ContextNotActiveException(RequestScoped.class.getName());
+        }
+        if (isActivator.compareAndSet(true, false)) {
+            requestContext.terminate();
+        }
+    }
+
+}

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/requestcontext/RequestContextTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/requestcontext/RequestContextTest.java
@@ -28,7 +28,6 @@ import io.quarkus.arc.ManagedContext;
 import io.quarkus.arc.test.ArcTestContainer;
 import javax.enterprise.context.ContextNotActiveException;
 import javax.enterprise.context.control.RequestContextController;
-
 import org.junit.Rule;
 import org.junit.Test;
 
@@ -83,7 +82,7 @@ public class RequestContextTest {
         assertNotEquals(controller2Id, client.getControllerId());
         assertTrue(Controller.DESTROYED.get());
     }
-    
+
     @Test
     public void testRequestContextController() {
         Controller.DESTROYED.set(false);

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/requestcontext/RequestContextTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/requestcontext/RequestContextTest.java
@@ -27,6 +27,8 @@ import io.quarkus.arc.ArcContainer;
 import io.quarkus.arc.ManagedContext;
 import io.quarkus.arc.test.ArcTestContainer;
 import javax.enterprise.context.ContextNotActiveException;
+import javax.enterprise.context.control.RequestContextController;
+
 import org.junit.Rule;
 import org.junit.Test;
 
@@ -37,7 +39,7 @@ public class RequestContextTest {
 
     @Test
     public void testRequestContext() {
-
+        Controller.DESTROYED.set(false);
         ArcContainer arc = Arc.container();
         ManagedContext requestContext = arc.requestContext();
 
@@ -73,6 +75,53 @@ public class RequestContextTest {
         requestContext.activate();
         assertNotEquals(controller2Id, arc.instance(Controller.class).get().getId());
         requestContext.terminate();
+        assertTrue(Controller.DESTROYED.get());
+
+        // @ActivateRequestContext
+        Controller.DESTROYED.set(false);
+        ControllerClient client = arc.instance(ControllerClient.class).get();
+        assertNotEquals(controller2Id, client.getControllerId());
+        assertTrue(Controller.DESTROYED.get());
+    }
+    
+    @Test
+    public void testRequestContextController() {
+        Controller.DESTROYED.set(false);
+        ArcContainer arc = Arc.container();
+        RequestContextController controller = Arc.container().instance(RequestContextController.class).get();
+
+        try {
+            arc.instance(Controller.class).get().getId();
+            fail();
+        } catch (ContextNotActiveException expected) {
+        }
+
+        controller.activate();
+        assertFalse(Controller.DESTROYED.get());
+        Controller controller1 = arc.instance(Controller.class).get();
+        Controller controller2 = arc.instance(Controller.class).get();
+        String controller2Id = controller2.getId();
+        assertEquals(controller1.getId(), controller2Id);
+        controller.deactivate();
+        assertTrue(Controller.DESTROYED.get());
+
+        try {
+            arc.instance(Controller.class).get().getId();
+            fail();
+        } catch (ContextNotActiveException expected) {
+        }
+
+        // Id must be different in a different request
+        Controller.DESTROYED.set(false);
+        controller.activate();
+        assertNotEquals(controller2Id, arc.instance(Controller.class).get().getId());
+        controller.deactivate();
+        assertTrue(Controller.DESTROYED.get());
+
+        Controller.DESTROYED.set(false);
+        controller.activate();
+        assertNotEquals(controller2Id, arc.instance(Controller.class).get().getId());
+        controller.deactivate();
         assertTrue(Controller.DESTROYED.get());
 
         // @ActivateRequestContext

--- a/test-framework/common/src/main/java/io/quarkus/test/common/PathTestHelper.java
+++ b/test-framework/common/src/main/java/io/quarkus/test/common/PathTestHelper.java
@@ -63,4 +63,23 @@ public final class PathTestHelper {
             return Paths.get(getTestClassesLocation(testClass).toString()
                     .replace(TEST_CLASSES_FRAGMENT_GRADLE, CLASSES_FRAGMENT_GRADLE));
     }
+
+    public static boolean isTestClass(String className, ClassLoader classLoader) {
+        String classFileName = className.replace('.', File.separatorChar) + ".class";
+        URL resource = classLoader.getResource(classFileName);
+        if (!resource.getProtocol().startsWith("file")) {
+            return false;
+        }
+        try {
+            Path path = Paths.get(resource.toURI());
+            if (path.toString().contains(TEST_CLASSES_FRAGMENT_MAVEN) ||
+                    path.toString().contains(TEST_CLASSES_FRAGMENT_GRADLE)) {
+                return true;
+            } else {
+                return false;
+            }
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+    }
 }

--- a/test-framework/common/src/main/java/io/quarkus/test/common/PathTestHelper.java
+++ b/test-framework/common/src/main/java/io/quarkus/test/common/PathTestHelper.java
@@ -67,17 +67,13 @@ public final class PathTestHelper {
     public static boolean isTestClass(String className, ClassLoader classLoader) {
         String classFileName = className.replace('.', File.separatorChar) + ".class";
         URL resource = classLoader.getResource(classFileName);
-        if (!resource.getProtocol().startsWith("file")) {
+        if (resource == null || !resource.getProtocol().startsWith("file")) {
             return false;
         }
         try {
             Path path = Paths.get(resource.toURI());
-            if (path.toString().contains(TEST_CLASSES_FRAGMENT_MAVEN) ||
-                    path.toString().contains(TEST_CLASSES_FRAGMENT_GRADLE)) {
-                return true;
-            } else {
-                return false;
-            }
+            return path.toString().contains(TEST_CLASSES_FRAGMENT_MAVEN) ||
+                    path.toString().contains(TEST_CLASSES_FRAGMENT_GRADLE);
         } catch (URISyntaxException e) {
             throw new RuntimeException(e);
         }

--- a/test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java
+++ b/test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java
@@ -49,7 +49,7 @@ public class TestResourceManager {
             try {
                 testResource.start();
             } catch (Exception e) {
-                throw new RuntimeException("Unable to start Quarkus test resource " + testResource);
+                throw new RuntimeException("Unable to start Quarkus test resource " + testResource, e);
             }
         }
     }
@@ -59,7 +59,7 @@ public class TestResourceManager {
             try {
                 testResource.stop();
             } catch (Exception e) {
-                throw new RuntimeException("Unable to start Quarkus test resource " + testResource);
+                throw new RuntimeException("Unable to start Quarkus test resource " + testResource, e);
             }
         }
     }

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
@@ -20,7 +20,6 @@ import static io.quarkus.test.common.PathTestHelper.getAppClassLocation;
 import static io.quarkus.test.common.PathTestHelper.getTestClassesLocation;
 
 import java.io.Closeable;
-import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
@@ -34,7 +33,12 @@ import java.util.List;
 import java.util.Map;
 import java.util.concurrent.LinkedBlockingDeque;
 import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
 
+import org.jboss.builder.BuildChainBuilder;
+import org.jboss.builder.BuildContext;
+import org.jboss.builder.BuildStep;
 import org.junit.jupiter.api.extension.AfterEachCallback;
 import org.junit.jupiter.api.extension.BeforeAllCallback;
 import org.junit.jupiter.api.extension.BeforeEachCallback;
@@ -48,11 +52,13 @@ import org.objectweb.asm.ClassWriter;
 
 import io.quarkus.deployment.ClassOutput;
 import io.quarkus.deployment.QuarkusClassWriter;
+import io.quarkus.deployment.builditem.TestClassPredicateBuildItem;
 import io.quarkus.deployment.util.IoUtil;
 import io.quarkus.runner.RuntimeRunner;
 import io.quarkus.runner.TransformerTarget;
 import io.quarkus.runtime.LaunchMode;
 import io.quarkus.test.common.NativeImageLauncher;
+import io.quarkus.test.common.PathTestHelper;
 import io.quarkus.test.common.PropertyTestUtil;
 import io.quarkus.test.common.RestAssuredURLManager;
 import io.quarkus.test.common.TestInjectionManager;
@@ -94,6 +100,8 @@ public class QuarkusTestExtension implements BeforeAllCallback, BeforeEachCallba
 
         Path appClassLocation = getAppClassLocation(context.getRequiredTestClass());
         Path testClassLocation = getTestClassesLocation(context.getRequiredTestClass());
+        ClassLoader testClassLoader = context.getRequiredTestClass().getClassLoader();
+
         RuntimeRunner runtimeRunner = RuntimeRunner.builder()
                 .setLaunchMode(LaunchMode.TEST)
                 .setClassLoader(getClass().getClassLoader())
@@ -190,6 +198,23 @@ public class QuarkusTestExtension implements BeforeAllCallback, BeforeEachCallba
                                 ex.printStackTrace();
                             }
                         }
+                    }
+                })
+                .addChainCustomizer(new Consumer<BuildChainBuilder>() {
+                    @Override
+                    public void accept(BuildChainBuilder buildChainBuilder) {
+                        buildChainBuilder.addBuildStep(new BuildStep() {
+                            @Override
+                            public void execute(BuildContext context) {
+                                context.produce(new TestClassPredicateBuildItem(new Predicate<String>() {
+                                    @Override
+                                    public boolean test(String className) {
+                                        return PathTestHelper.isTestClass(className, testClassLoader);
+                                    }
+                                }));
+                            }
+                        }).produces(TestClassPredicateBuildItem.class)
+                                .build();
                     }
                 })
                 .build();

